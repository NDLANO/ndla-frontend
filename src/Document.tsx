/**
 * Copyright (c) 2025-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ReactNode } from "react";
import type { ManifestChunk } from "vite";
import Scripts from "./components/Scripts/Scripts";
import config from "./config";

interface Props {
  language: string;
  children?: ReactNode;
  chunks?: ManifestChunk[];
  devEntrypoint: string;
}

const getUniqueCss = (chunks: ManifestChunk[]) => {
  const uniq = chunks.reduce((acc, curr) => {
    curr.css?.forEach((css) => acc.add(css));
    return acc;
  }, new Set<string>());
  return Array.from(uniq);
};

export const Document = ({ language, children, chunks = [], devEntrypoint }: Props) => {
  const faviconEnvironment = config.ndlaEnvironment === "dev" ? "test" : config.ndlaEnvironment;
  const locale = language === "nb" || language === "nn" ? "no" : language;

  const [entryPoint, ...importedChunks] = chunks;
  const css = getUniqueCss(chunks);

  return (
    <html lang={locale}>
      <head>
        <link rel="icon" type="image/png" sizes="32x32" href={`/static/favicon-${faviconEnvironment}-32x32.png`} />
        <link rel="icon" type="image/png" sizes="16x16" href={`/static/favicon-${faviconEnvironment}-16x16.png`} />
        <link
          rel="preload"
          href={`/locales/${language}/translation.json`}
          as="fetch"
          type="application/json"
          crossOrigin="anonymous"
        />
        <link
          rel="apple-touch-icon"
          type="image/png"
          sizes="180x180"
          href={`/static/apple-touch-icon-${faviconEnvironment}.png`}
        />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1 viewport-fit=cover" />
        <link href="https://api.fontshare.com/v2/css?f[]=satoshi@1&display=swap" rel="stylesheet" />
      </head>
      <body>
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={{
            __html: `
      window.dataLayer = window.dataLayer || [];
      window._mtm = window._mtm || [];
      window.originalLocation = {
        originalLocation:
          document.location.protocol +
          "//" +
          document.location.hostname +
          document.location.pathname +
          document.location.search,
      };
      window.dataLayer.push(window.originalLocation);
`,
          }}
        ></script>
        {config.runtimeType === "development" && (
          <>
            <script
              dangerouslySetInnerHTML={{
                __html: `
                import RefreshRuntime from "/@react-refresh";
                RefreshRuntime.injectIntoGlobalHook(window);
                window.$RefreshReg$ = () => {};
                window.$RefreshSig$ = () => (type) => type;
                window.__vite_plugin_react_preamble_installed__ = true;
              `,
              }}
              type="module"
            />
            <script src="/@vite/client" type="module" />
            <script type="module" src={`/${devEntrypoint}`}></script>
          </>
        )}
        <script
          // We're hydrating the entire document. Our config differentiates between server and client, so it's necessary to suppress any hydration warnings here. TODO: Find a better workaround for this
          suppressHydrationWarning
          dangerouslySetInnerHTML={{
            __html: config.isClient ? "" : `window.DATA = "$WINDOW_DATA"`,
          }}
        ></script>
        <Scripts />
        {!!entryPoint && <script type="module" src={`/${entryPoint.file}`}></script>}
        {css.map((file) => (
          <link rel="stylesheet" href={`/${file}`} key={file} />
        ))}
        {importedChunks.map((chunk) => (
          <link rel="modulepreload" href={`/${chunk.file}`} key={chunk.file}></link>
        ))}
        <div id="root">{children}</div>
      </body>
    </html>
  );
};
