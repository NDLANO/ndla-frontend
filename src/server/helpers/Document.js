/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import serialize from 'serialize-javascript';
import ScriptLoader from '@ndla/polyfill/lib/ScriptLoader';
import { GoogleTagMangerScript, GoogleTagMangerNoScript } from './Gtm';
import config from '../../config';

const Document = ({ helmet, assets, data, css, ids }) => {
  const htmlAttrs = helmet.htmlAttributes.toComponent();
  const bodyAttrs = helmet.bodyAttributes.toComponent();

  return (
    <html {...htmlAttrs}>
      <head>
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet="utf-8" />
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1 viewport-fit=cover"
        />
        <link
          rel="stylesheet"
          href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,300i,400,400i,600,700|Source+Serif+Pro:400,700|Source+Code+Pro:400,700"
        />
        {config.gaTrackingId && (
          <script async src="https://www.google-analytics.com/analytics.js" />
        )}
        <GoogleTagMangerScript />
        {helmet.title.toComponent()}
        {helmet.meta.toComponent()}
        {helmet.link.toComponent()}
        {assets.css && <link rel="stylesheet" href={assets.css} />}
        <link
          rel="shortcut icon"
          href="/static/ndla-favicon.png"
          type="image/x-icon"
        />
        {css && ids && (
          <style data-emotion-css={`${ids.join(' ')}`}>${css}</style>
        )}
      </head>
      <body {...bodyAttrs}>
        <GoogleTagMangerNoScript />
        <script
          dangerouslySetInnerHTML={{
            __html: `
            window.dataLayer = window.dataLayer || [];
            window.originalLocation = { originalLocation: document.location.protocol + '//' + document.location.hostname + document.location.pathname + document.location.search };
            window.dataLayer.push(window.originalLocation);`,
          }}
        />
        <div id="root">REPLACE_ME</div>
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={{
            __html: `window.DATA = ${serialize(data)}; `,
          }}
        />
        <ScriptLoader polyfill={assets.polyfill} scripts={assets.js} />
        {helmet.script.toComponent()}
      </body>
    </html>
  );
};

Document.propTypes = {
  helmet: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  data: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  assets: PropTypes.shape({
    css: PropTypes.string,
    js: PropTypes.array.isRequired,
    polyfill: PropTypes.shape({
      src: PropTypes.oneOfType([PropTypes.string, PropTypes.array]).isRequired,
    }),
  }).isRequired,
  css: PropTypes.string,
  ids: PropTypes.arrayOf(PropTypes.string),
};

export default Document;
