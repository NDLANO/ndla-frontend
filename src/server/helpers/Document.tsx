/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { HelmetServerState } from 'react-helmet-async';
import serialize from 'serialize-javascript';
import ScriptLoader from '@ndla/polyfill/lib/ScriptLoader';
import { GoogleTagMangerScript, GoogleTagMangerNoScript } from './Gtm';
import { Matomo } from './Matomo';
import Tagmanager from './Tagmanager';
import config, { ConfigType } from '../../config';

export interface Assets {
  css?: string;
  js: { src: string }[];
  mathJaxConfig?: { js: string };
  polyfill?: { src: string };
}

export interface DocumentData {
  initialProps?: any;
  apolloState?: any;
  assets: Assets;
  config?: ConfigType;
}

interface Props {
  helmet: HelmetServerState;
  assets: Assets;
  data?: DocumentData;
  styles?: string;
}

const Document = ({ helmet, assets, data, styles }: Props) => {
  const htmlAttrs = helmet.htmlAttributes.toComponent();
  const bodyAttrs = helmet.bodyAttributes.toComponent();

  return (
    // eslint-disable-next-line jsx-a11y/html-has-lang
    <html {...htmlAttrs}>
      <head>
        <Tagmanager />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet="utf-8" />
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1 viewport-fit=cover"
        />
        {config.gaTrackingId ? (
          <script async src="https://www.google-analytics.com/analytics.js" />
        ) : null}
        <GoogleTagMangerScript />
        {helmet.title.toComponent()}
        {helmet.meta.toComponent()}
        {helmet.link.toComponent()}
        {assets.css && <link rel="stylesheet" href={assets.css} />}
        <link
          rel="shortcut icon"
          href="/static/ndla-favicon.png"
          type="image/x-icon"
        />
        {helmet.script.toComponent()}
        {styles && <div dangerouslySetInnerHTML={{ __html: styles }} />}
      </head>
      <body {...bodyAttrs}>
        <GoogleTagMangerNoScript />
        <Matomo />
        <script
          dangerouslySetInnerHTML={{
            __html: `
            window.dataLayer = window.dataLayer || [];
            window.originalLocation = { originalLocation: document.location.protocol + '//' + document.location.hostname + document.location.pathname + document.location.search };
            window.dataLayer.push(window.originalLocation);`,
          }}
        />
        <div id="root">REPLACE_ME</div>
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={{
            __html: `window.DATA = ${serialize(data)}; `,
          }}
        />
        <ScriptLoader polyfill={assets.polyfill} scripts={assets.js} />
      </body>
    </html>
  );
};

export default Document;
