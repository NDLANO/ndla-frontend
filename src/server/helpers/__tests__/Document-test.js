/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Helmet, HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { PageContainer } from '@ndla/ui';
import Document from '../Document';

HelmetProvider.canUseDOM = false;

test('Document renderers correctly', () => {
  const helmetContext = {};
  // Render page with Helmet component
  render(
    <HelmetProvider context={helmetContext}>
      <PageContainer>
        <h1>Hello World</h1>
        <Helmet
          htmlAttributes={{ lang: 'nb' }}
          title="NDLA test title"
          meta={[{ name: 'description', content: 'NDLA meta description' }]}
        />
      </PageContainer>
    </HelmetProvider>,
  );

  // Create and render Document. Match snapshot of rendered document.
  const { container } = render(
    <Document
      locale="nb"
      data={{}}
      helmet={helmetContext.helmet}
      assets={{
        css: '/main.css',
        js: [{ src: '/client.js' }, { src: '/vendor.js' }],
      }}
    />,
  );
  const header = container.querySelector('html');

  // Match snapshot for rendered Document. Should contain title, html lang,
  // and meta description from PageContainer > Helmet.
  expect(header).toMatchSnapshot();
});
