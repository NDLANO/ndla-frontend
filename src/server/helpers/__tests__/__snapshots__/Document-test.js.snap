// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document renderers correctly 1`] = `
<html
  lang="nb"
>
  <head>
    <meta
      content="IE=edge"
      httpEquiv="X-UA-Compatible"
    />
    <meta
      charSet="utf-8"
    />
    <meta
      content="width=device-width, initial-scale=1 viewport-fit=cover"
      name="viewport"
    />
    <link
      href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,300i,400,400i,600,700|Source+Serif+Pro:400,700|Source+Code+Pro:400,700"
      rel="stylesheet"
    />
    
    <title
      data-rh={true}
    >
      NDLA test title
    </title>
    <meta
      content="NDLA meta description"
      data-rh={true}
      name="description"
    />
    <link
      href="/main.css"
      rel="stylesheet"
    />
    <link
      href="/static/ndla-favicon.png"
      rel="shortcut icon"
      type="image/x-icon"
    />
  </head>
  <body>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
            window.dataLayer = window.dataLayer || [];
            window.originalLocation = { originalLocation: document.location.protocol + '//' + document.location.hostname + document.location.pathname + document.location.search };
            window.dataLayer.push(window.originalLocation);",
        }
      }
    />
    <div
      id="root"
    >
      REPLACE_ME
    </div>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "window.DATA = {}; ",
        }
      }
      type="text/javascript"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
  var feats = [];
  ('fetch' in window) || feats.push('fetch');('includes' in Array.prototype) || feats.push('Array.prototype.includes');('fill' in Array.prototype) || feats.push('Array.prototype.fill');('open' in document.createElement('details')) || feats.push('details');('assign' in Object) || feats.push('Object.assign');('remove' in Element.prototype) || feats.push('Element.prototype.remove');('includes' in String.prototype) || feats.push('String.prototype.includes');('startsWith' in String.prototype) || feats.push('String.prototype.startsWith');('endsWith' in String.prototype) || feats.push('String.prototype.endsWith');('isInteger' in Number) || feats.push('Number.isInteger');('isNaN' in Number) || feats.push('Number.isNaN');

  function cs(script, useHead, onLoad) {
    var s = document.createElement('script');
    s.src = script.src;
    s.async = true;
    s.crossorigin = 'anonymous';
    if (onLoad) { s.onload = onLoad; }
    if (useHead) {
      document.head.appendChild(s);
    } else {
      document.body.appendChild(s);
    }
  }

  function _polyfillComplete() {
    var scripts = [{\\"src\\":\\"/client.js\\"},{\\"src\\":\\"/vendor.js\\"}];
    for (var i = 0; i < scripts.length; i++) {
      cs(scripts[i], false);
    }
  }

  if (feats.length && {\\"src\\":\\"/polyfill.js\\"}) {
    cs({\\"src\\":\\"/polyfill.js\\"}, true, _polyfillComplete);
  } else {
    _polyfillComplete();
  }
  ",
        }
      }
      type="text/javascript"
    />
  </body>
</html>
`;
